### ISSUE TITLE
Add a trim option to a chart to drop unused columns in a dataframe

### ISSUE THREAD

**Issue by govinda-kamath** [altair-viz/altair#3013]
Hello, 

Do you think it would be possible to add a `trim` function to a chart or to the `to_json` function of a chart. This seems to be just getting all `field` from `chart.encoding` and restricting the `chart.data` dataframe to the columns there. My use-case for this is when I export a bunch of altair plots to a html. Having to keep track of columns used restrict the dataframe to those columns ends up being quite annoying and conducive to bug. 

Example with current behaviour:
```
import altair as alt
from vega_datasets import data
import json

source = data.cars().iloc[:2,:]

chart = alt.Chart(source).mark_circle(size=60).encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color='Origin',
    tooltip=['Name', 'Horsepower', 'Miles_per_Gallon']
).interactive()

json.loads(chart.to_json())
```
gives me 
```
{'$schema': 'https://vega.github.io/schema/vega-lite/v4.17.0.json',
 'config': {'view': {'continuousHeight': 300, 'continuousWidth': 400}},
 'data': {'name': 'data-6135d1d5ab873ca6aa6011d3e6597f3a'},
 'datasets': {'data-6135d1d5ab873ca6aa6011d3e6597f3a': [{'Acceleration': 12.0,
    'Cylinders': 8,
    'Displacement': 307.0,
    'Horsepower': 130.0,
    'Miles_per_Gallon': 18.0,
    'Name': 'chevrolet chevelle malibu',
    'Origin': 'USA',
    'Weight_in_lbs': 3504,
    'Year': '1970-01-01T00:00:00'},
   {'Acceleration': 11.5,
    'Cylinders': 8,
    'Displacement': 350.0,
    'Horsepower': 165.0,
    'Miles_per_Gallon': 15.0,
    'Name': 'buick skylark 320',
    'Origin': 'USA',
    'Weight_in_lbs': 3693,
    'Year': '1970-01-01T00:00:00'}]},
 'encoding': {'color': {'field': 'Origin', 'type': 'nominal'},
  'tooltip': [{'field': 'Name', 'type': 'nominal'},
   {'field': 'Horsepower', 'type': 'quantitative'},
   {'field': 'Miles_per_Gallon', 'type': 'quantitative'}],
  'x': {'field': 'Horsepower', 'type': 'quantitative'},
  'y': {'field': 'Miles_per_Gallon', 'type': 'quantitative'}},
 'mark': {'size': 60, 'type': 'circle'},
 'selection': {'selector003': {'bind': 'scales',
   'encodings': ['x', 'y'],
   'type': 'interval'}}}
```
which includes unused columns like `Acceleration`, `Weight_in_lbs`, etc. I was thinking of writing in an option so that the behaviour would be 
```
json.loads(chart.to_json(trim=True))
```
which outputs
```
{'$schema': 'https://vega.github.io/schema/vega-lite/v4.17.0.json',
 'config': {'view': {'continuousHeight': 300, 'continuousWidth': 400}},
 'data': {'name': 'data-49be283546a66f7642a4f34bac266124'},
 'datasets': {'data-49be283546a66f7642a4f34bac266124': [{'Horsepower': 130.0,
    'Miles_per_Gallon': 18.0,
    'Name': 'chevrolet chevelle malibu',
    'Origin': 'USA'},
   {'Horsepower': 165.0,
    'Miles_per_Gallon': 15.0,
    'Name': 'buick skylark 320',
    'Origin': 'USA'}]},
 'encoding': {'color': {'field': 'Origin', 'type': 'nominal'},
  'tooltip': [{'field': 'Name', 'type': 'nominal'},
   {'field': 'Horsepower', 'type': 'quantitative'},
   {'field': 'Miles_per_Gallon', 'type': 'quantitative'}],
  'x': {'field': 'Horsepower', 'type': 'quantitative'},
  'y': {'field': 'Miles_per_Gallon', 'type': 'quantitative'}},
 'mark': {'size': 60, 'type': 'circle'},
 'selection': {'selector004': {'bind': 'scales',
   'encodings': ['x', 'y'],
   'type': 'interval'}}}
```

I was thinking of just going over all fields in the `chart.encoding` and restricting the dataframe to the fields seen there if `trim` is set before writing it out. Do you think this is reasonable, or are there cases where this will break things?

---

Please follow these steps to make it more efficient to respond to your feature request.

- [x] Since Altair is a Python wrapper around the Vega-Lite visualization grammar, [most feature requests should be reported directly to Vega-Lite](https://github.com/vega/vega-lite/issues). You can click the Action Button of your Altair chart and "Open in Vega Editor" to create a reproducible Vega-Lite example.
- [x] Search for duplicate issues.
- [x] Describe the feature's goal, motivating use cases, and its expected behavior.

**Comment by joelostblom** [comment]
Thanks for you post @govinda-kamath , I am going to close this as a duplicate of https://github.com/altair-viz/altair/issues/2428. Please continue the discussion there and note that it is possible to achieve what you want currently if you use Vega Fusion https://vegafusion.io/:

> The VegaFusion mime renderer can be used to overcome the Altair [MaxRowsError](https://altair-viz.github.io/user_guide/faq.html#maxrowserror-how-can-i-plot-large-datasets) by performing data-intensive aggregations on the server **and pruning unused columns from the source dataset**. First install the vegafusion Python package with the embed extras enabled

**Comment by jonmmease** [comment]
Hi @govinda-kamath, the [VegaFusion Mime Renderer](https://vegafusion.io/mime_renderer.html) has the ability to trim columns and pre-apply Vega transforms before embedding the result in the Jupyter notebook.  We don't currently expose a function to do this when saving to an HTML file, but it's a good idea and I opened https://github.com/hex-inc/vegafusion/issues/292 to track it.

**Comment by jonmmease** [comment]
@govinda-kamath, this functionality is now available in VegaFusion 1.2. See https://vegafusion.io/save.html
